### PR TITLE
Add FuHidDevice abstraction

### PR DESCRIPTION
**Pull request by hughsie**
Quite a few plugins use HID commands to communicate with the hardware. At the
moment we have ~6 implementations of SET_REPORT and are soon to add one more.

Move this into common code.

### CODE REVIEW

File: plugins/dell-dock/fu-dell-dock-hid.c [hunk 1]
```diff
@@ -81,17 +81,14 @@ fu_dell_dock_hid_set_report (FuDevice *self,
...
-	gsize actual_len = 0;
 	for (gint i = 1; i <= HID_MAX_RETRIES; i++) {
 		g_autoptr(GError) error_local = NULL;
-		ret = g_usb_device_control_transfer (
-		    usb_device, G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
-		    G_USB_DEVICE_REQUEST_TYPE_CLASS,
-		    G_USB_DEVICE_RECIPIENT_INTERFACE, FU_HID_REPORT_SET, 0x0200,
-		    0x0000, outbuffer, 192, &actual_len,
-		    HIDI2C_TRANSACTION_TIMEOUT, NULL, &error_local);
+		ret = fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0,
```
Comment (superm1) [inline plugins/dell-dock/fu-dell-dock-hid.c]: I think that 0x0 should actually be 0x200 here, no?

File: plugins/dell-dock/fu-dell-dock-hid.c [hunk 2]
```diff
@@ -121,17 +113,14 @@ fu_dell_dock_hid_get_report (FuDevice *self,
...
-	gsize actual_len = 0;
 	for (gint i = 1; i <= HID_MAX_RETRIES; i++) {
 		g_autoptr(GError) error_local = NULL;
-		ret = g_usb_device_control_transfer (
-		    usb_device, G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
-		    G_USB_DEVICE_REQUEST_TYPE_CLASS,
-		    G_USB_DEVICE_RECIPIENT_INTERFACE, FU_HID_REPORT_GET, 0x0100,
-		    0x0000, inbuffer, 192, &actual_len,
-		    HIDI2C_TRANSACTION_TIMEOUT, NULL, &error_local);
+		ret = fu_hid_device_get_report (FU_HID_DEVICE (self), 0x0,
```
Comment (superm1) [inline plugins/dell-dock/fu-dell-dock-hid.c]: and this one should be 0x100 right?

File: plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c [hunk 3]
```diff
@@ -342,18 +343,11 @@ fu_logitech_hidpp_bootloader_request (FuLogitechHidPpBootloader *self,
...
-						    G_USB_DEVICE_RECIPIENT_INTERFACE,
-						    FU_HID_REPORT_SET,
-						    0x0200, 0x0000,
-						    buf_request,
-						    sizeof (buf_request),
-						    &actual_length,
-						    FU_UNIFYING_DEVICE_TIMEOUT_MS,
-						    NULL,
-						    error)) {
+		if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0,
```
Comment (superm1) [inline plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c]: houldn't this also be 0x200?

File: plugins/rts54hid/fu-rts54hid-device.c [hunk 4]
```diff
@@ -101,7 +43,10 @@ fu_rts54hid_device_set_clock_mode (FuRts54HidDevice *self, gboolean enable, GErr
 	};
 	guint8 buf[FU_RTS54FU_HID_REPORT_LENGTH] = { 0 };
 	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
-	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error)) {
+	if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0, buf, sizeof(buf),
```
Comment (superm1) [inline plugins/rts54hid/fu-rts54hid-device.c]: this one should also be 0x200

File: plugins/rts54hid/fu-rts54hid-device.c [hunk 5]
```diff
@@ -120,7 +65,10 @@ fu_rts54hid_device_reset_to_flash (FuRts54HidDevice *self, GError **error)
...
```
Comment (superm1) [inline plugins/rts54hid/fu-rts54hid-device.c]: 0x200 for this one?

File: plugins/rts54hid/fu-rts54hid-device.c [hunk 6]
```diff
@@ -176,10 +127,16 @@ fu_rts54hid_device_verify_update_fw (FuRts54HidDevice *self, GError **error)
...
 	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
-	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error))
+	if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0, buf, sizeof(buf),
+				       FU_RTS54HID_DEVICE_TIMEOUT * 2,
+				       FU_HID_DEVICE_FLAG_NONE,
+				       error))
 		return FALSE;
 	g_usleep (4 * G_USEC_PER_SEC);
-	if (!fu_rts54hid_device_get_report (self, buf, sizeof(buf), error))
+	if (!fu_hid_device_get_report (FU_HID_DEVICE (self), 0x0, buf, sizeof(buf),
```
Comment (superm1) [inline plugins/rts54hid/fu-rts54hid-device.c]: 0x100 for this one

File: plugins/rts54hid/fu-rts54hid-device.c [hunk 7]
```diff
@@ -176,10 +127,16 @@ fu_rts54hid_device_verify_update_fw (FuRts54HidDevice *self, GError **error)
 
 	/* set then get */
 	memcpy (buf, &cmd_buffer, sizeof(cmd_buffer));
-	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error))
+	if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0, buf, sizeof(buf),
```
Comment (superm1) [inline plugins/rts54hid/fu-rts54hid-device.c]: 0x200 for this one

File: plugins/rts54hid/fu-rts54hid-device.c [hunk 8]
```diff
@@ -152,7 +100,10 @@ fu_rts54hid_device_write_flash (FuRts54HidDevice *self,
 			     data, data_sz, 0x0,					/* src */
 			     data_sz, error))
 		return FALSE;
-	if (!fu_rts54hid_device_set_report (self, buf, sizeof(buf), error)) {
+	if (!fu_hid_device_set_report (FU_HID_DEVICE (self), 0x0, buf, sizeof(buf),
```
Comment (superm1) [inline plugins/rts54hid/fu-rts54hid-device.c]: 0x200 for this one

File: plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c (hunk 3 above)
Comment (hughsie) [reply plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c]: The wValue is being set in fu_hid_device_set_report - so 0x0200 is `FU_HID_REPORT_TYPE_OUTPUT << 8`

File: plugins/synaptics-cxaudio/fu-synaptics-cxaudio-device.c [hunk 9]
```diff
@@ -88,30 +64,11 @@ fu_synaptics_cxaudio_device_output_report (FuSynapticsCxaudioDevice *self,
...
-			     (guint) actual_length, bufsz);
-		return FALSE;
-	}
-
-	/* success */
-	return TRUE;
+	return fu_hid_device_set_report (FU_HID_DEVICE (self), buf[0],
+					 buf, bufsz,
+					 FU_SYNAPTICS_CXAUDIO_USB_TIMEOUT,
+					 FU_HID_DEVICE_FLAG_NONE,
```
Comment (superm1) [inline plugins/synaptics-cxaudio/fu-synaptics-cxaudio-device.c]: I think you're missing a flag here

File: plugins/synaptics-cxaudio/fu-synaptics-cxaudio-device.c [hunk 10]
```diff
@@ -121,37 +78,11 @@ fu_synaptics_cxaudio_device_input_report (FuSynapticsCxaudioDevice *self,
...
-			     (guint) actual_length, bufsz);
-		return FALSE;
-	}
-
-	/* success */
-	return TRUE;
+	return fu_hid_device_get_report (FU_HID_DEVICE (self), ReportID,
+					 buf, bufsz,
+					 FU_SYNAPTICS_CXAUDIO_USB_TIMEOUT,
+					 FU_HID_DEVICE_FLAG_NONE,
```
Comment (superm1) [inline plugins/synaptics-cxaudio/fu-synaptics-cxaudio-device.c]: I think also missing a flag here

File: plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c (hunk 3 above)
Comment (superm1) [reply plugins/logitech-hidpp/fu-logitech-hidpp-bootloader.c]: oh... I completely missed that at the top of the function.  OK never mind, I guess most of my comments are wrong then :)

**Comment by superm1** [comment]
The CI failure is this: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954146, a fix is pending so you can merge anyway.

**Pull request merged**